apigee/registry
# Build the Implementation+Plumbing for a Sample Plugin

## Summary
We would like to define a way for users to create their own plugins. This follows a plugin pattern that follows a pattern previously introduced by `protoc` and `gnostic`. The idea is that a plugin is simply a standalone binary executable, which takes in a request through `Stdin` and returns a response on `Stdout`. The request and response [schema are well defined](https://github.com/apigee/registry/pull/334). The conformance command will then run these plugins, where the name of each plugin is simply `registry-lint-LINTERNAME`.

In this pull request, I introduce a `sample` plugin which can be used as a building block for other plugins that we will introduced (namely Spectral and API Linter). I also modify the conformance command so that it takes in an extra flag, `plugin`, which acts as a signal to tell the command to run using plugins rather than our previous approach.

## Note
I have not modified the old code, but rather added new code to define the behavior of the conformance command with plugins, which can be enabled with `--plugin=true`. The idea is that once all is set in stone and the Spectral + API Linter plugins work correctly, we can get rid of the old code completely.

## Test
- Have registry running on `localhost:8080`.
- Ensure that https://github.com/APIs-guru/openapi-directory is extracted to your Desktop.
- Ensure that the following style guide is located under `~/styleguide.yaml`:
```yaml
id: samplestyleguide
mime_types:
  - application/x.openapi+gzip;version=2
  - application/x.openapi+gzip;version=3
guidelines:
  - id: sampleguideline
    display_name: Sample Guideline to test linter protos
    description: This guideline governs properties for ref fields on specs.
    rules:
      - id: sample1
        description: This is the first sample rule.
        linter: sample
        linter_rulename: sample1
        severity: ERROR
      - id: sample2
        description: This is the second sample rule.
        linter: sample
        linter_rulename: sample2
        severity: WARNING
    status: 2
linters:
  - name: sample
```
- Compile the registry project by running `make` in the root directory
- Change directory to `$GOPATH/bin`
- Now, run the following series of bash commands:
```bash
apg registry delete-project --name projects/example

apg registry create-project --project_id example --insecure --address localhost:8080

registry upload bulk openapi --project-id example ~/Desktop/openapi-directory/APIs

registry upload styleguide ~/styleguide-sample.yaml --project-id=example

registry compute conformance projects/example/locations/global/apis/zappiti.com/versions/4.15.174/specs/swagger.yaml --plugin=true

registry get projects/example/locations/global/apis/zappiti.com/versions/4.15.174/specs/swagger.yaml/artifacts/conformance-samplestyleguide --contents
```
Ensure that the output matches the [following gist](https://gist.github.com/muhammadharis/82911d9052e0a9164037f8459ffe3514).
